scd/token.c

```diff
@@ -585,6 +659,50 @@ token_new(const token_constr_data_t *constr_data)
 		break;
 	}
 #endif // ENABLESCHSM
+#ifdef ENABLEPKCS11
+	case (PKCS11): {
+		DEBUG("Create scd_token with internal type 'PKCS11'");
+
+		ASSERT(constr_data->uuid);
+		ASSERT(constr_data->init_str.pkcs11_module);
+		p11token_t *token;
+		token = p11token_token_by_label(constr_data->init_str.pkcs11_module,
+						constr_data->uuid);
+
+		// create new token if not found
+		if (token == NULL) {
+			// create random so_pin
+			unsigned char *random_mem =
+				(unsigned char *)file_read_new("/dev/urandom", 32);
+			ASSERT(random_mem);
+			str_t *so_pin = str_hexdump_new(random_mem, 32);
+			free(random_mem);
+			token = p11token_create_p11(constr_data->init_str.pkcs11_module,
+						    str_buffer(so_pin), STOKEN_DEFAULT_PASS,
+						    constr_data->uuid);
+			if (!token) {
+				ERROR("Creation of softtoken failed");
+				goto err;
+			}
+			str_free(so_pin, true);
+		}
+
+		new_token->token_data->int_token.p11token = token;
+
+		new_token->token_data->type = PKCS11;
+		new_token->lock = int_lock_p11;
+		new_token->unlock = int_unlock_p11;
+		new_token->is_locked = int_is_locked_p11;
+		new_token->is_locked_till_reboot = int_is_locked_till_reboot_p11;
+		new_token->wrap_key = int_wrap_key_p11;
+		new_token->unwrap_key = int_unwrap_key_p11;
+		new_token->change_passphrase = int_change_passphrase_p11;
+		new_token->reset_auth = int_reset_auth_p11;
+		new_token->get_atr = int_get_atr_p11;
+		new_token->send_apdu = int_send_apdu_p11;
+		break;
```

Comment:
For a followup refactoring, we should consider to introduce a struct for all those functions and hide
the token specific functions in the corresponding p11token.c/usbtoken.c files (analogues to the compartment_module_t)